solidity/contracts/isms/native/OptimismIsm.sol

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+import "forge-std/console.sol";
+
+import {IInterchainSecurityModule} from "../../interfaces/IInterchainSecurityModule.sol";
+import {IOptimismMessageHook} from "../../interfaces/hooks/IOptimismMessageHook.sol";
+import {Message} from "../../libs/Message.sol";
+import {TypeCasts} from "../../libs/TypeCasts.sol";
+
+import {ICrossDomainMessenger} from "@eth-optimism/contracts/libraries/bridge/ICrossDomainMessenger.sol";
+import {CrossDomainOwnable3} from "@eth-optimism/contracts-bedrock/contracts/L2/CrossDomainOwnable3.sol";
+
+contract OptimismIsm is IInterchainSecurityModule, CrossDomainOwnable3 {
+    mapping(bytes32 => bool) public receivedEmitters;
+
+    event ReceivedMessage(bytes32 indexed messageId, address indexed emitter);
+
+    constructor(ICrossDomainMessenger _l2Messenger) {
+        l2Messenger = _l2Messenger;
+    }
+
+    ICrossDomainMessenger public l2Messenger;
+    IOptimismMessageHook public hook;
+
+    // solhint-disable-next-line const-name-snakecase
+    uint8 public constant moduleType =
+        uint8(IInterchainSecurityModule.Types.NATIVE);
+
+    function setOptimismHook(IOptimismMessageHook _hook) external {
```

Comment:
This needs authentication. Alternatively, consider putting it in the constructor and making it immutable

Reply:
Ah I guess there's a chicken and egg problem, where they both need to know about eachother, and so making them immutable is tricky.

Technically it's possible since we can predict the address of one or the other, but maybe that's not worth it.

If you don't want to force deployers to do that (which I suggest you do not), you might consider prioritizing L1 gas (i.e. make the ism address immutable on the hook, and make the hook address initializable on the ISM)